scripts/cameo-tag.py

```diff
@@ -0,0 +1,115 @@
+# partition we must have the tags in place. To solve this we insert the
+# tag two times into the kernel image.
+#
+# To understand what we do here it is helpful to explain how the original 
+# CAMEO tag generation/checking works. The firmware consists of two parts.
+# A kernel uImage (<1.5MB) and a rootfs image (<12MB) that are written to 
+# their respective mtd partitions. The default generator simply checksums
+# both parts and appends 16 bytes [<CAMEOTAG><0001><checksum>] to each part. 
+# The checksum is only an addition of all preceding bytes (b0+b1+b2+...). 
+# A tag does not interfere with any data in the images itself. During boot
+# the loader will scan all primary/secondary partitions (2*kernel, 2*rootfs) 
+# until it finds the CAMEO tag. If checksums match everything is fine. 
+# If all 4 fail we are lost. Luckily the loader does not care about where
+# the tags are located and ignores any data beyond a tag.
+#
+# The OpenWRT image consists of a kernel (>1.5MB) and a rootfs. There is
+# no chance to add CAMEO tags at the default locations. This would leave 
+# the kernel partition without a tag. So we must find suitable space.
+# 
+# Location for original kernel partition is at the end of the uImage header. 
+# We will reuse the last bytes of the IH_NAME field. This is the tricky 
+# part because we have the header CRC and the CAMEO checksum that must match
+# the whole header. uImage header CRC checksums all data expect the CRC data
+# itself. The same applies for CAMEO checksum. It checksums all data except
+# itself. Changing one of both results in a change of the other. So we will
+# brute force checksum combinations until we get an exact match between 
+# those two. This method usually converges after ~500 iterations (<0.2s) and 
+# is easier than using reverse math. After manipulation the uImage header
+# looks like this:
+#
+# [....<header_crc>....<shortened_name><00><CAMEOTAG><0001><checksum>] 
+#
+# Location for original rootfs partition is very simple. It is behind the 
+# OpenWRT compressed kernel image file that spans into the rootfs. So
+# the tag will be written somewhere to the following rootfs partition and 
+# can be found by U-Boot. The CAMEO checksum calculation must start at
+# position 0x180000 (see DLINK_KERNEL_PART_SIZE) and includes the "second"
+# half of the "split" kernel uImage. 
+
+import argparse
+import pathlib
+import socket
+import struct
+import zlib
+
+UIMAGE_HEADER = 64
+CAMEO_TAG = bytearray([0x43, 0x41, 0x4d, 0x45, 0x4f, 0x54, 0x41, 0x47, 0x00, 0x00, 0x00, 0x01])
+NO_CRC = bytearray([0, 0, 0, 0])
+
+def read_buffer(offset, whence, count):
+    args.uimage_file.seek(offset, whence)
+    return bytearray(args.uimage_file.read(count))
+
+def write_buffer(offset, whence, buf):
+    args.uimage_file.seek(offset, whence)
+    args.uimage_file.write(buf)
+
+def toarray(value):
+    return value.to_bytes(4, 'big')
+
+def cameosum(array):
+    return sum(iter(array)) % (1<<32)
```

Comment:
If this is called with something that isn't an iterable, this probably should complain anyway. So I think `sum(array) % (1<<32)` is sufficient.

Reply:
fixed